CASSANDRA-16524 Allow for setting buffer max capacity.

grighetto opened the pull request:
CLOSED IN FAVOR OF #6

vice committed:
Fix init script for debian Buster

Patch by Vicente Jimenez Aguilar, reviewed by brandonwilliams for
CASSANDRA-15770

polo-language committed:
Eliminate gen-doc template warning and unused (problematic) import

gen-doc is now python3 compatible

 patch by Angelo Polo; reviewed by Ekaterina Dimitrova, Mick Semb Wever for CASSANDRA-16206

bereng committed:
CASSANDRA-16532 Fix flaky testSkipScrubCorruptedCounterRowWithTool

patch by Berenguer Blasi; reviewed by Andrés de la Peña for CASSANDRA-16532

blerer committed:
Fix Streaming Repair metrics

patch by Benjamin Lerer; reviewed by Ekaterina Dimitrova for
CASSANDRA-16190

michaelsembwever committed:
Migrate dependency handling from maven-ant-tasks to resolver-ant-tasks

- This removes lib/ from version control.
- The generated pom files are now the source of truth to the project's depdendencies, and the lib/ folder re-created during the build from compile scope dependencies excluding provided and system scopes. Project dependencies are first downloaded to the maven local repository at `~/.m2/repository`.
- Maven is now required to be installed to build, that is the `mvn` command needs to be found in the path.
- The new resolver-ant-tasks targets have been encapsulated into the `build/build-resolver.xml` ant file.
- Most classpaths in build.xml are replaced with resolver:resolve definitions.
- The build-deps-pom generated pom file is now only used for declaring test dependencies.

 patch by Mick Semb Wever; reviewed by Brandon Williams for CASSANDRA-16391

grighetto committed:
CASSANDRA-16524 Allow for setting buffer max capacity.

grighetto committed:
CASSANDRA-16524 Allow for setting buffer max capacity.

grighetto committed:
Use MIDRES config for CircleCI build.

grighetto committed:
Use MIDRES config for CircleCI build.

grighetto committed:
CASSANDRA-16524 Ensure a buffer's maxCapacity is at least the size of its initial capacity.

grighetto committed:
CASSANDRA-16524 Ensure a buffer's maxCapacity is at least the size of its initial capacity.

grighetto committed:
CASSANDRA-16524 Added unit tests to ensure a buffer's max capacity is respected.

grighetto committed:
CASSANDRA-16524 Added unit tests to ensure a buffer's max capacity is respected.

jasonstack commented on test/unit/org/apache/cassandra/utils/memory/BufferPoolTest.java:
```diff
@@ -146,6 +149,32 @@ public void testRecycle()
+    public void testBufferDefaultMaxCapacity()
+    {
+        DatabaseDescriptor.clientInitialization();
+        ByteBuf noMaxCapacity = GlobalBufferPoolAllocator.instance.buffer(100);
```
nit: would be nice to release the buffer

grighetto replied:
Done

ekaterinadimitrova2 replied:
Shouldn't we release it also at other places?

jasonstack commented on src/java/org/apache/cassandra/net/GlobalBufferPoolAllocator.java:
```diff
@@ -36,6 +36,6 @@ private GlobalBufferPoolAllocator()
 
     static ByteBuf wrap(ByteBuffer buffer)
```
the use cases of this method are fixed-length buffer. should be fine to add `maxCapacity`

grighetto replied:
@jasonstack How about using `buffer.capacity()` here then?
Since `maxCapacity` is not ignored anymore, I had to come up with a default value for this method, so I went with netty's default (Integer.MAX_VALUE) to be on the safe side, but if the use case here is fixed-length buffers and this is not an overridden method (it's only used internally), I guess we can go with `buffer.capacity()`.
If you think that's reasonable, I can push that change and run another build in CircleCI.

The main issue was in `BufferPoolAllocator#newDirectBuffer` which was called by netty but the maxCapacity param was ignored.

jasonstack replied:
work for me. thanks

grighetto replied:
@jasonstack Change pushed. Running new CI builds.

jasonstack commented on src/java/org/apache/cassandra/net/GlobalBufferPoolAllocator.java:
```diff
@@ -36,6 +36,6 @@ private GlobalBufferPoolAllocator()
     static ByteBuf wrap(ByteBuffer buffer)
     {
-        return new Wrapped(instance, buffer);
+        return new Wrapped(instance, buffer, Integer.MAX_VALUE);
```
nit: do you mind adding a comment saying this is the same as `AbstractByteBufAllocator#DEFAULT_MAX_CAPACITY`?

michaelsembwever committed:
Offline build mode

When all dependencies are already in the local `~/.m2/repository/` we should not be needing an internet connection. ("Offline" mode should be automatic)

 patch by Mick Semb Wever; reviewed by Brandon Williams for CASSANDRA-16559

michaelsembwever committed:
Update IDE, Eclipse and NetBeans, classpaths for tests

 patch by Mick Semb Wever, Berenguer Blasi; reviewd by Ekaterina Dimitrova, Benjamin Lerer for CASSANDRA-16560

grighetto committed:
CASSSANDRA-16524 Release buffer in unit tests (per PR feedback).

grighetto committed:
CASSSANDRA-16524 Release buffer in unit tests (per PR feedback).

yifan-c committed:
Fix race between secondary index building and active compactions tracking

patch by Yifan Cai; reviewed by Andres de la Peña, Jon Meredith for CASSANDRA-16554

ekaterinadimitrova2 committed:
Scheduled (delayed) schema pull tasks should not run after MIGRATION stage shutdown during decommission
patch by Ekaterina Dimitrova; reviewed by Caleb Rackliffe and Yifan Cai for CASSANDRA-16495

krummas committed:
Make sure sstables with moved starts are removed correctly in LeveledGenerations

Patch by marcuse; reviewed by Benedict Elliott Smith for CASSANDRA-16552

bereng commented:
LGTM pending CI

grighetto committed:
CASSANDRA-16524 Use `buffer.capacity()` as max capacity for fixed-length buffers in `GlobalBufferPoolAllocator#wrap`.

grighetto committed:
CASSANDRA-16524 Use `buffer.capacity()` as max capacity for fixed-length buffers in `GlobalBufferPoolAllocator#wrap`.

krummas committed:
Add JStackJUnitTask to avoid downloading the jar

Patch by marcuse; reviewed by Brandon Williams, David Capwell, Michael Semb Wever for CASSANDRA-16570

michaelsembwever committed:
Merge branch 'cassandra-3.11' into trunk

dcapwell committed:
When behind a firewall trunk is not buildable, need to allow overriding URLs

patch by David Capwell; reviewed by Brandon Williams, Jon Meredith, Michael Semb Wever for CASSANDRA-16563

ekaterinadimitrova2 committed:
Merge branch 'cassandra-3.0' into cassandra-3.11

ekaterinadimitrova2 committed:
Merge branch 'cassandra-3.11' into trunk

ekaterinadimitrova2 committed:
Fix warning about multiple versions of ant detected when running ant test
patch by Ekaterina Dimitrova; reviewed by Brandon Williams and Mick Semb Wever for CASSANDRA-16571

michaelsembwever committed:
Fix dtest-jar building when switching branches in circleci
patch by Brandon Williams and Mick Semb Wever; tested and reviewed by Ekaterina Dimitrova for CASSANDRA-16572

ekaterinadimitrova2 committed:
Merge branch 'cassandra-3.0' into cassandra-3.11

ekaterinadimitrova2 committed:
Merge branch 'cassandra-3.11' into trunk

michaelsembwever committed:
Don't put apidocs (javadoc) into the binary artifact

 patch by Mick Semb Wever; reviewed by Brandon Williams for CASSANDRA-15561

bereng committed:
Flaky StorageServiceServerTest

patch by Berenguer Blasi; reviewed by Michael Semb Wever for CASSANDRA-16569

michaelsembwever committed:
Don't fork jvms in the build

On build machines with less than 1GB ram, use ANT_OPTS="-Xmx512m"

 patch by Mick Semb Wever; reviewed by Brandon Williams for CASSANDRA-16583

michaelsembwever committed:
Upgrade jflex to 1.8.2

 patch by Mick Semb Wever; reviewed by Brandon Williams for CASSANDRA-16576

driftx committed:
Don't wait for migrations from removed nodes, mention flag to skip

Patch by brandonwilliams; reviewed by Adam Holmberg, adelapena and
bdeggleston for CASSANDRA-16577

driftx committed:
Don't wait for migrations from removed nodes, mention flag to skip

Patch by brandonwilliams; reviewed by Adam Holmberg, adelapena and
bdeggleston for CASSANDRA-16577

driftx committed:
Don't wait for migrations from removed nodes, mention flag to skip

Patch by brandonwilliams; reviewed by Adam Holmberg, adelapena and
bdeggleston for CASSANDRA-16577

michaelsembwever committed:
Remove test parallelism from ant build.xml

 patch by Mick Semb Wever; reviewed by Berenguer Blasi, Yifan Cai for CASSANDRA-16595

grighetto committed:
CASSANDRA-16524 Return wrapped buffer to pool when resizing; Added more tests around buffer allocation.

grighetto committed:
CASSANDRA-16524 Return wrapped buffer to pool when resizing; Added more tests around buffer allocation.

grighetto committed:
CASSANDRA-16524 Prevent OOME in buffer pool unit test.

grighetto committed:
CASSANDRA-16524 Prevent OOME in buffer pool unit test.

driftx committed:
Fix mixed cluster GROUP BY queries

Patch by Benjamin Lerer and Adam Holmberg, reviewed by brandonwilliams
for CASSANDRA-16582

Co-authored-by: Benjamin Lerer <blerer@apache.org>
Co-authored-by: Adam Holmberg <adam.holmberg@datastax.com>

Commit:
Add autocomplete, errors for provide_overlapping_tombstones

Patch by Brian Houser; reviewed by marcuse for CASSANDRA-16350

bereng commented on src/java/org/apache/cassandra/net/BufferPoolAllocator.java:
```diff
@@ -94,17 +98,37 @@ void release()
+            if (newCapacity == capacity())
+                return this;
+
+            // resizing doesn't use the pool
```
I don't understand why the pool isn't used here. You mean you've checked the code for `UnpooledUnsafeDirectByteBuf`and it doesn't use the pool upon resizing?

grighetto replied:
Initially, I noticed it doesn't use the pool because releasing a resized buffer made the pool's occupancy go negative.
In other words, calling `GlobalBufferPoolAllocator.instance.usedSizeInBytes()` after `resizedBuffer.release()` would return a negative value.

And I confirmed that by looking at netty's source too:

https://github.com/netty/netty/blob/netty-4.1.58.Final/buffer/src/main/java/io/netty/buffer/UnpooledDirectByteBuf.java#L156

https://github.com/netty/netty/blob/netty-4.1.58.Final/buffer/src/main/java/io/netty/buffer/UnpooledDirectByteBuf.java#L104

It calls `ByteBuffer.allocateDirect` and doesn't use the internal allocator we pass in the constructor, which explains the behavior I described above. Since we're extending **Unpooled**UnsafeDirectByteBuf, I guess it makes sense.

The check I added around `returnToPool` prevents putting a ByteBuffer back into the pool that hasn't originated in the pool.

bereng replied:
OK yes I was referring to that, whether you had checked we were keeping it consistent with netty.

bereng commented on test/unit/org/apache/cassandra/utils/memory/BufferPoolTest.java:
```diff
@@ -154,19 +155,98 @@ public void testBufferDefaultMaxCapacity()
+
+        byte[] bufferContent = new byte[300];
+
+        BufferPoolAllocator.Wrapped wrapped = (BufferPoolAllocator.Wrapped)buffer;
```
space `(BufferPoolAllocator.Wrapped)buffer` -> `(BufferPoolAllocator.Wrapped) buffer`

grighetto replied:
Fixed.

krummas committed:
Merge branch 'cassandra-3.11' into trunk

driftx committed:
Merge branch 'cassandra-3.11' into trunk

dcapwell committed:
Periodic failures in *RepairCoordinator*Test caused by race condition with nodetool repair

patch by David Capwell; reviewed by Marcus Eriksson for CASSANDRA-16585

driftx committed:
Merge branch 'cassandra-3.0' into cassandra-3.11

driftx committed:
Merge branch 'cassandra-3.11' into trunk

grighetto committed:
CASSANDRA-16524 Fixed minor code formatting issue per PR feedback.

grighetto committed:
CASSANDRA-16524 Fixed minor code formatting issue per PR feedback.

yifan-c committed:
Fix false unavailable for queries due to cluster topology changes

patch by Yifan Cai; reviewed by Aleksey Yeschenko, Andres de la Peña for CASSANDRA-16545

michaelsembwever committed:
Merge branch 'cassandra-2.2' into cassandra-3.0

michaelsembwever committed:
Merge branch 'cassandra-3.0' into cassandra-3.11

michaelsembwever committed:
Merge branch 'cassandra-3.11' into trunk

michaelsembwever committed:
Restore running each test class in its own separate jvm and cassandra directory
 (Remove test parallelism from ant build.xml)

 patch by Mick Semb Wever; reviewed by Yifan Cai, David Capwell for CASSANDRA-16595

dcapwell committed:
Use source release of python driver from pip rather than GitHub

patch by David Capwell; reviewed by Adam Holmberg, Michael Semb Wever for CASSANDRA-16599

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/net/BufferPoolAllocator.java:
```diff
@@ -76,11 +79,14 @@ void put(ByteBuffer buffer)
         bufferPool.putUnusedPortion(buffer);
     }
 
+    public long usedSizeInBytes() { return bufferPool.usedSizeInBytes(); }
```
@VisibleForTesting is missing

michaelsembwever committed:
Merge branch 'cassandra-2.2' into cassandra-3.0

michaelsembwever committed:
Merge branch 'cassandra-3.0' into cassandra-3.11

michaelsembwever committed:
Merge branch 'cassandra-3.11' into trunk

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/net/BufferPoolAllocator.java:
```diff
@@ -76,11 +79,14 @@ void put(ByteBuffer buffer)
 
-    void putUnusedPortion(ByteBuffer buffer)
+    @VisibleForTesting
+    public void putUnusedPortion(ByteBuffer buffer)
```
Do we really want this method to get public?

jasonstack commented on src/java/org/apache/cassandra/net/BufferPoolAllocator.java:
```diff
@@ -92,17 +98,37 @@ void release()
+                return this;
+
+            // resizing doesn't use the pool
+            ByteBuf newBuffer = super.capacity(newCapacity);
```
not sure why we need to use `super.capacity(newCapacity);` which doesn't use our buffer pool. 

is it possible to directly allocate new capacity from buffer pool? and then copy old buffer to new buffer. at the end, release the old buffer. So we don't have to keep track of `returnToPool` which defeats the purpose of reducing fragmented buffer allocation.

wdyt? @grighetto @bereng

grighetto replied:
@jasonstack I went that route initially, but there's a problem with that approach, we pass the initial buffer to the constructor of `UnpooledUnsafeDirectByteBuf` and it keeps a reference to that internally:

https://github.com/netty/netty/blob/netty-4.1.58.Final/buffer/src/main/java/io/netty/buffer/UnpooledDirectByteBuf.java#L42

We would have to call `setByteBuffer` to replace that object, but it is not public:

https://github.com/netty/netty/blob/netty-4.1.58.Final/buffer/src/main/java/io/netty/buffer/UnpooledDirectByteBuf.java#L114

If we don't replace that internal buffer, there will be mismatches because some internal functions use it.
`super.capacity(newCapacity)` already does the right things, creates a new buffer, copies over the bytes, releases the old one and replaces that reference.

> So we don't have to keep track of returnToPool which defeats the purpose of reducing fragmented buffer allocation

Right, I thought about that too and I agree, but for most cases it will behave as before (it has been working with fixed-length buffers up until now). Increasing the buffer capacity should be an edge case as the one reported in the ticket.

bereng replied:
That is the worry I had and I raised in my previous comment. You have to check Netty code to see if everything is consistent with Netty.

bereng replied:
> Increasing the buffer capacity should be an edge case as the one reported in the ticket

I think that's the key here. Making pooled, unpooled and netty work together nicely seems to be the problem.

BTW @grighetto did you check if upon resizing we should be checking any pool/general config memory size limit?

jasonstack replied:
one more issues with current implementation: when `capacity` is called, the `newBuffer` is never deallocated on buf `release` unless `capacity` is called again. But again, the latest `newBuffer` with the latest capacity is not deallocated.

jasonstack replied:
the `newBuffer` created by `super.capacity()` is not freed by `freeDirect()`, would be nice to handle it. the rest looks good

ekaterinadimitrova2 replied:
I have the same questions as @bereng here. I want to do this to match the 3.11 implementation if possible with minimum hurdle.
Not saying @grighetto didn't do it, he did a lot and great job, I am just confirming those details.

ekaterinadimitrova2 replied:
If in 3.11 it is always used by Netty Integer.MAX_VALUE I would go for that path probably. Confirming with @grighetto the details in Slack

driftx committed:
Merge branch 'cassandra-2.2' into cassandra-3.0

driftx committed:
Merge branch 'cassandra-3.0' into cassandra-3.11

driftx committed:
Merge branch 'cassandra-3.11' into trunk

grighetto committed:
CASSANDRA-16524 Moved buffer allocation tests to the o.a.c.net package; Added overloaded method for `putUnusedPortion` that takes a `BufferPoolAllocator.Wrapped` object; Updated callers of `Wrapped.adopt()` to put unused bytes before adopting.

grighetto committed:
CASSANDRA-16524 Moved buffer allocation tests to the o.a.c.net package; Added overloaded method for `putUnusedPortion` that takes a `BufferPoolAllocator.Wrapped` object; Updated callers of `Wrapped.adopt()` to put unused bytes before adopting.

grighetto committed:
Merge branch '16524-allow-for-increasing-buffer-capacity' of https://github.com/grighetto/cassandra into 16524-allow-for-increasing-buffer-capacity

# Conflicts:
#	.circleci/config.yml
#	CHANGES.txt
